### PR TITLE
Fix testing workflow

### PR DESCRIPTION
**Pull request by bquan0**
This pull request is a continuation of #13 and #16. It changes `testing.yml` to make it run the testing workflow on one of the four docker images that corresponds to the python version being used in the workflow.

### CODE REVIEW

**Comment by bquan0** [comment]
The testing workflow still runs into the "missing GLIBC_2.29" error for python 3.8, which I don't know how to fix. This error also sometimes cancels the workflows for the other versions and prevents them from finishing.

**Comment by gonuke** [comment]
I think I figured out the problem - and it was probably bad advice from me a few weeks ago.

When I look at the `buildImage` workflow, it configures the local runner with the correct python version, but then it builds a docker image from the `Dockerfile`.  The local python configuration is irrelevant! 

Moreover, you build the docker image from the `python:3.8-slim-buster` image and then use it for every version of python.

We also install all the dependencies in the Docker image, by design, but all are the 3.8 version.

I'll have to think about the best solution if we do want to test against different python versions.  We may need to install the dependencies during the workflow rather than in the docker image.

**Comment by gonuke** [comment]
Another option - we can use the python version as a docker `build_arg` and start from a different base image with each of the docker images we build.

**Comment by bquan0** [comment]
I think I was able to get the docker images to build by using the `build_arg` solution you proposed. I added a `PY_VERSION` arg to the `Dockerfile` and made the `Dockerfile` build from the python image with the input python version. I also made sure to pass in the `PY_VERSION` argument in `buildImage.yml` in the `Build and push Docker image` step.

Let me know if there's any issues with the changes. Hopefully this will solve the GLIBC_2.29 error when it's pushed to the main branch.

File: .github/workflows/buildImage.yml [hunk 1]
```diff
@@ -38,6 +38,8 @@ jobs:
         uses: docker/build-push-action@v2
         with:
           context: .
+          build-args: |
+            PY_VERSION=${{ matrix.python-version }}
           push: true
           tags: ${{ env.REGISTRY }}/${{ github.repository }}/learn-maintenance-image:latest${{ matrix.python-version }}
```
Comment (gonuke) [inline .github/workflows/buildImage.yml:44]: Let's put the python version in the image name and not the version tag:
```suggestion
          tags: ${{ env.REGISTRY }}/${{ github.repository }}/learn-maintenance-image-${{ matrix.python-version }}:latest
```

File: requirements.txt [hunk 2]
```diff
@@ -4,5 +4,5 @@ itsdangerous==1.1.0
 Jinja2==2.11.3
 MarkupSafe==1.1.1
 Werkzeug==1.0.1
-numpy==1.23.1
+numpy==1.21.6
```
Comment (gonuke) [inline requirements.txt:7]: Why do we use an older version of NUMPY?

File: .github/workflows/testing.yml [hunk 3]
```diff
@@ -14,8 +25,6 @@ jobs:
         uses: actions/setup-python@v4
         with:
           python-version: ${{ matrix.python-version }}
```
Comment (gonuke) [inline .github/workflows/testing.yml:27]: Things I don't know: Do we still need to setup the python version if we are using images with specific python versions?

**Review by gonuke — CHANGES_REQUESTED** [review]
This looks a lot more promising - thanks @bquan0

File: requirements.txt (hunk 2 above)
Comment (bquan0) [reply requirements.txt]: The suggestion to change `numpy==1.21.6` to just `numpy` worked. Now all the docker images download the most recent version of numpy for the respective python version.

**Comment by gonuke** [comment]
This needs an update to the `buildImage` to tag images in a way that matches this preferred choice.

**Comment by gonuke** [comment]
Let's PR the build image change alone to get the images updated and then confirm the testing in a separate PR

**Comment by bquan0** [comment]
Just reverted the changes to `testing.yml`. I think this pull request is ready to be merged. I'll open a new pull request with the changes to `testing.yml`.

**Review by gonuke — APPROVED** [review]

**Pull request merged**